includes/classes/shopping_cart.php

```diff
@@ -989,62 +1038,75 @@ public function attributes_price($product_id)
                 continue;
             }
 
-            $this->notify('NOTIFY_CART_ATTRIBUTES_PRICE_NEXT', $product_id, $attribute_price->fields);
+            $this->notify('NOTIFY_CART_ATTRIBUTES_PRICE_NEXT', $uprid, $attribute_price->fields);
 
             $new_attributes_price = 0;
             $discount_type_id = '';
             $sale_maker_discount = '';
 
-            if ($attribute_price->fields['product_attribute_is_free'] == '1' && zen_get_products_price_is_free((int)$product_id)) {
+            if ($attribute_price->fields['product_attribute_is_free'] === '1' && zen_get_products_price_is_free($prid)) {
                 // no charge
             } else {
                 // + or blank adds
-                if ($attribute_price->fields['price_prefix'] == '-') {
+                if ($attribute_price->fields['price_prefix'] === '-') {
                     // calculate proper discount for attributes
-                    if ($attribute_price->fields['attributes_discounted'] == '1') {
+                    if ($attribute_price->fields['attributes_discounted'] === '1') {
                         $discount_type_id = '';
                         $sale_maker_discount = '';
-                        $new_attributes_price = zen_get_discount_calc($product_id, $attribute_price->fields['products_attributes_id'], $attribute_price->fields['options_values_price'], $qty);
-                        $attributes_price -= ($new_attributes_price);
+                        $new_attributes_price = zen_get_discount_calc($prid, $attribute_price->fields['products_attributes_id'], $attribute_price->fields['options_values_price'], $qty);
+                        $attributes_price -= $new_attributes_price;
                     } else {
                         $attributes_price -= $attribute_price->fields['options_values_price'];
                     }
+                } elseif ($attribute_price->fields['attributes_discounted'] === '1') {
+                    // calculate proper discount for attributes
+                    $discount_type_id = '';
+                    $sale_maker_discount = '';
+                    $new_attributes_price = zen_get_discount_calc(
+                        $prid,
+                        $attribute_price->fields['products_attributes_id'],
+                        $attribute_price->fields['options_values_price'] +
+                            (zen_get_products_price_is_priced_by_attributes($attribute_price->fields['products_id']) ?
+                                zen_products_lookup($attribute_price->fields['products_id'], 'products_price') : 0.0),
+                        $qty
+                    );
+                    $new_attributes_price -= (zen_get_products_price_is_priced_by_attributes($prid)) ? zen_products_lookup($prid, 'products_price') : 0;
```

Comment:
I'd rather put the closing `)` at the end of the line, instead of at the end of the function call. Or skip the wrapping parentheses.

Reply:
Noting that this section has changed "a fair amount" for #6017 (sigh).